src/Raven.Server/Documents/Handlers/AI/Agents/AbstractAiAgentProcessor.cs

```diff
@@ -189,31 +203,129 @@ public void ValidateForResume()
+
+                    await HandleQueryToolCallsAsync(context, configuration, document, aiResponse);
+
+                    if (TryGetUserTools(context, document, configuration, aiResponse))
+                        break; // we need to return the user tool requests to the client, so we can continue the conversation
+                }
+            }
+
+            var history = await TryReduceChatSize();
+
+            return (aiResponse.Result, document, history);
 
-            while (true)
+
+            async Task<BlittableJsonReaderObject> TryReduceChatSize()
             {
-                aiResponse = await client.CompleteAsync(
-                    context,
-                    document.Messages,
-                    tools,
-                    document.TotalUsage,
-                    token
-                );
-                if (aiResponse.Type is AiResponseType.Result)
-                    break;
-
-                await HandleQueryToolCallsAsync(context, configuration, document, aiResponse);
-
-                if (TryGetUserTools(context, document, configuration, aiResponse))
-                    break; // we need to return the user tool requests to the client, so we can continue the conversation
+                var reduction = configuration.ChatReduction;
+                if (reduction == null || document.OpenActionCalls.Count > 0)
+                    return null;
+
+                var clone = reduction.History == null ? null : document.ToBlittable(context, configuration, reduction.History?.HistoryExpiration);
+
+                if (reduction.Truncate != null)
+                {
+                    if (document.Messages.Count > reduction.Truncate.MessagesLengthBeforeTruncate)
+                    {
+                        var truncateCount = document.Messages.Count - reduction.Truncate.MessagesLengthAfterTruncate;
+                        truncateCount = int.Min(truncateCount, document.Messages.Count - 1); // prevent System.ArgumentException (out of bounds)
+                        if(truncateCount > 0)
+                            document.Messages.RemoveRange(1, truncateCount);
+                    }
+                }
+                else if (reduction.Tokens != null)
+                {
+                    if (aiUsage.TotalTokens > reduction.Tokens.MaxTokensBeforeSummarization)
+                        await SummarizeAsync(context, conStr, reduction.Tokens, document, token);
+                }
+
+                return clone;
             }
+        }
 
-            return (aiResponse.Result, document);
+        private async Task SummarizeAsync(JsonOperationContext context, AiConnectionString connectionString, AiAgentSummarizationByTokens summarization, ConversationDocument oldChat, CancellationToken token)
+        {
+            var systemPrompt = oldChat.Messages.FirstOrDefault();
+            if (systemPrompt == null)
+                throw new InvalidOperationException($"System prompt cannot be null.");
```

Comment:
Explain which system prompt you are talking about, and in what context this operation failed

Reply:
Done here:
https://github.com/ravendb/ravendb/pull/20973